AsyncDisplayKit/Details/ASCollectionDataController.mm

```diff
@@ -0,0 +1,217 @@
+/* Copyright (c) 2015-present, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ */
+
+#import "ASCollectionDataController.h"
+
+#import "ASAssert.h"
+#import "ASMultidimensionalArrayUtils.h"
+#import "ASDisplayNode.h"
+#import "ASDisplayNodeInternal.h"
+#import "ASDataController+Subclasses.h"
+
+//#define LOG(...) NSLog(__VA_ARGS__)
+#define LOG(...)
+
+@interface ASCollectionDataController ()
+
+- (id<ASCollectionDataControllerSource>)collectionDataSource;
+
+@end
+
+@implementation ASCollectionDataController {
+  NSMutableDictionary *_pendingNodes;
+  NSMutableDictionary *_pendingIndexPaths;
+}
+
+- (void)prepareForReloadData
+{
+  _pendingNodes = [NSMutableDictionary dictionary];
+  _pendingIndexPaths = [NSMutableDictionary dictionary];
+
+  [[self supplementaryKinds] enumerateObjectsUsingBlock:^(NSString *kind, NSUInteger idx, BOOL *stop) {
+    LOG(@"Populating elements of kind: %@", kind);
+    NSMutableArray *indexPaths = [NSMutableArray array];
```

Comment:
Please make make the dictionaries and arrays created in this file use a capacity-aware initializer, e.g. arrayWithCapacity, wherever possible.